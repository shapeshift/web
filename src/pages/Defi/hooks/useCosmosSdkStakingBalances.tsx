import type { AccountId, AssetId, ChainId } from '@shapeshiftoss/caip'
import { useMemo } from 'react'
import type { BigNumber } from 'lib/bignumber/bignumber'
import { bn, bnOrZero } from 'lib/bignumber/bignumber'
import type { ActiveStakingOpportunity } from 'state/slices/selectors'
import {
  selectAssetById,
  selectMarketDataById,
  selectStakingOpportunitiesDataFullByFilter,
} from 'state/slices/selectors'
import { useAppSelector } from 'state/store'

type UseCosmosStakingBalancesProps = {
  accountId?: AccountId | undefined
  assetId: AssetId
}

export type UseCosmosStakingBalancesReturn = {
  cosmosSdkStakingOpportunities: MergedActiveStakingOpportunity[]
  totalBalance: string
}

export type MergedActiveStakingOpportunity = ActiveStakingOpportunity & {
  fiatAmount?: string
  assetId: AssetId
  chainId: ChainId
  tvl: string
  isLoaded?: boolean
}

export function useCosmosSdkStakingBalances({
  assetId,
  accountId,
}: UseCosmosStakingBalancesProps): UseCosmosStakingBalancesReturn {
  const marketData = useAppSelector(state => selectMarketDataById(state, assetId))
  const asset = useAppSelector(state => selectAssetById(state, assetId))

  const filter = useMemo(
    () => ({
      accountId: accountId ?? '',
      assetId,
    }),
    [accountId, assetId],
  )
  const stakingOpportunities = useAppSelector(s =>
    selectStakingOpportunitiesDataFullByFilter(s, filter),
  )

  const mergedActiveStakingOpportunities = useMemo(() => {
    if (!asset) return []
    if (!marketData?.price) return []

    return Object.values(stakingOpportunities).map(opportunity => {
      const fiatAmount = bnOrZero(opportunity.totalDelegations)
        .div(`1e+${asset.precision}`)
        .times(bnOrZero(marketData.price))
        .toFixed(2)

      const tvl = bnOrZero(opportunity.tokens)
        .div(`1e+${asset.precision}`)
        .times(bnOrZero(marketData?.price))
        .toString()
      const data = {
        ...opportunity,
        cryptoAmount: bnOrZero(opportunity.totalDelegations)
          .div(`1e+${asset?.precision}`)
          .decimalPlaces(asset.precision)
          .toString(),
        tvl,
        fiatAmount,
        chainId: asset.chainId,
        assetId,
      }
      return data
    })
  }, [stakingOpportunities, assetId, asset, marketData])

  const totalBalance = useMemo(
    () =>
      Object.values(mergedActiveStakingOpportunities).reduce(
        (acc: BigNumber, opportunity: MergedActiveStakingOpportunity) => {
          return acc.plus(bnOrZero(opportunity.fiatAmount))
        },
        bn(0),
      ),
    [mergedActiveStakingOpportunities],
  )

  return {
    cosmosSdkStakingOpportunities: mergedActiveStakingOpportunities,
    totalBalance: totalBalance.toString(),
  }
}
