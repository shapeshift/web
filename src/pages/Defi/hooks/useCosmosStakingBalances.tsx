import { AssetId } from '@shapeshiftoss/caip'
import { chainAdapters, ChainTypes } from '@shapeshiftoss/types'
import { useMemo } from 'react'
import { BigNumber, bn, bnOrZero } from 'lib/bignumber/bignumber'
import {
  ActiveStakingOpportunity,
  selectAssetById,
  selectFirstAccountSpecifierByChainId,
  selectMarketDataById,
  selectStakingOpportunitiesDataFull,
} from 'state/slices/selectors'
import { useAppSelector } from 'state/store'

type UseCosmosStakingBalancesProps = {
  assetId: AssetId
}

export type UseCosmosStakingBalancesReturn = {
  cosmosStakingOpportunities: MergedActiveStakingOpportunity[]
  totalBalance: string
}

export type MergedActiveStakingOpportunity = ActiveStakingOpportunity & {
  fiatAmount?: string
  tokenAddress: string
  assetId: AssetId
  chain: ChainTypes
  tvl: string
  isLoaded?: boolean
}

export type MergedStakingOpportunity = chainAdapters.cosmos.Validator & {
  tokenAddress: string
  assetId: AssetId
  chain: ChainTypes
  tvl: string
}

export function useCosmosStakingBalances({
  assetId,
}: UseCosmosStakingBalancesProps): UseCosmosStakingBalancesReturn {
  const marketData = useAppSelector(state => selectMarketDataById(state, assetId))
  const asset = useAppSelector(state => selectAssetById(state, assetId))

  const accountSpecifier = useAppSelector(state =>
    selectFirstAccountSpecifierByChainId(state, asset?.chainId),
  )

  const stakingOpportunities = useAppSelector(state =>
    selectStakingOpportunitiesDataFull(state, { accountSpecifier, assetId }),
  )

  const mergedActiveStakingOpportunities = useMemo(() => {
    if (!marketData?.price) return []

    return Object.values(stakingOpportunities).map(opportunity => {
      const fiatAmount = bnOrZero(opportunity.totalDelegations)
        .div(`1e+${asset.precision}`)
        .times(bnOrZero(marketData.price))
        .toFixed(2)

      const tvl = bnOrZero(opportunity.tokens)
        .div(`1e+${asset.precision}`)
        .times(bnOrZero(marketData?.price))
        .toString()
      const data = {
        ...opportunity,
        cryptoAmount: bnOrZero(opportunity.totalDelegations)
          .div(`1e+${asset?.precision}`)
          .decimalPlaces(asset.precision)
          .toString(),
        tvl,
        fiatAmount,
        chain: asset.chain,
        assetId,
        tokenAddress: asset.slip44.toString(),
      }
      return data
    })
  }, [stakingOpportunities, assetId, asset, marketData])

  const totalBalance = useMemo(
    () =>
      Object.values(mergedActiveStakingOpportunities).reduce(
        (acc: BigNumber, opportunity: MergedActiveStakingOpportunity) => {
          return acc.plus(bnOrZero(opportunity.fiatAmount))
        },
        bn(0),
      ),
    [mergedActiveStakingOpportunities],
  )

  return {
    cosmosStakingOpportunities: mergedActiveStakingOpportunities,
    totalBalance: totalBalance.toString(),
  }
}
