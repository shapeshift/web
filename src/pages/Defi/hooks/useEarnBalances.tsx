import { bnOrZero } from 'lib/bignumber/bignumber'

import { useVaultBalances, UseVaultBalancesReturn } from './useVaultBalances'

export type UseEarnBalancesReturn = {
  vaults: UseVaultBalancesReturn
  totalEarningBalance: string
}

export function useEarnBalances(): UseEarnBalancesReturn {
  const vaults = useVaultBalances()
  // When staking, farming, lp, etc are added sum up the balances here
  const totalEarningBalance = bnOrZero(vaults?.totalBalance).toString()
  return { vaults, totalEarningBalance }
}
