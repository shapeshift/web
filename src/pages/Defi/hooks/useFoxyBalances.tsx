import { AssetId, toAssetId } from '@shapeshiftoss/caip'
import { DefiType, FoxyApi, WithdrawInfo } from '@shapeshiftoss/investor-foxy'
import { ChainTypes, NetworkTypes } from '@shapeshiftoss/types'
import { getConfig } from 'config'
import { useFoxy } from 'features/defi/contexts/FoxyProvider/FoxyProvider'
import { useCallback, useEffect, useMemo, useState } from 'react'
import { useSelector } from 'react-redux'
import { useChainAdapters } from 'context/PluginProvider/PluginProvider'
import { useWallet } from 'hooks/useWallet/useWallet'
import { BigNumber, bn, bnOrZero } from 'lib/bignumber/bignumber'
import { PortfolioBalancesById } from 'state/slices/portfolioSlice/portfolioSliceCommon'
import {
  selectAssets,
  selectMarketData,
  selectPortfolioAssetBalances,
  selectPortfolioLoading,
} from 'state/slices/selectors'

export type FoxyOpportunity = {
  type: DefiType
  provider: string
  version: string
  contractAddress: string
  rewardToken: string
  stakingToken: string
  chain: ChainTypes
  tvl?: BigNumber
  expired?: boolean
  apy?: string
  balance: string
  contractAssetId: AssetId
  tokenAssetId: AssetId
  rewardTokenAssetId: AssetId
  pricePerShare: BigNumber
  withdrawInfo: WithdrawInfo
}

export type MergedFoxyOpportunity = FoxyOpportunity & {
  cryptoAmount: string
  fiatAmount: string
}
export type UseFoxyBalancesReturn = {
  opportunities: MergedFoxyOpportunity[]
  totalBalance: string
  loading: boolean
}

async function getFoxyOpportunities(
  balances: PortfolioBalancesById,
  api: FoxyApi,
  userAddress: string,
) {
  const acc: Record<string, FoxyOpportunity> = {}
  const opps = await api.getFoxyOpportunities()
  for (let index = 0; index < opps.length; index++) {
    // TODO: assetIds in vaults
    const opportunity = opps[index]
    const withdrawInfo = await api.getWithdrawInfo({
      contractAddress: opportunity.contractAddress,
      userAddress,
    })
    const rewardTokenAssetId = toAssetId({
      chain: opportunity.chain,
      network: NetworkTypes.MAINNET,
      assetNamespace: 'erc20',
      assetReference: opportunity.rewardToken,
    })
    const contractAssetId = toAssetId({
      chain: opportunity.chain,
      network: NetworkTypes.MAINNET,
      assetNamespace: 'erc20',
      assetReference: opportunity.contractAddress,
    })
    const tokenAssetId = toAssetId({
      chain: opportunity.chain,
      network: NetworkTypes.MAINNET,
      assetNamespace: 'erc20',
      assetReference: opportunity.stakingToken,
    })
    const balance = balances[rewardTokenAssetId]

    const pricePerShare = api.pricePerShare()
    acc[opportunity.contractAddress] = {
      ...opportunity,
      balance: bnOrZero(balance).toString(),
      contractAssetId,
      tokenAssetId,
      rewardTokenAssetId,
      pricePerShare: bnOrZero(pricePerShare),
      withdrawInfo,
    }
  }
  return acc
}

export function useFoxyBalances(): UseFoxyBalancesReturn {
  const [loading, setLoading] = useState(false)
  const [opportunities, setOpportunites] = useState<Record<string, FoxyOpportunity>>({})
  const marketData = useSelector(selectMarketData)
  const assets = useSelector(selectAssets)

  const chainAdapterManager = useChainAdapters()

  const {
    state: { wallet },
  } = useWallet()

  const { foxy, loading: foxyLoading } = useFoxy()
  const balances = useSelector(selectPortfolioAssetBalances)
  const balancesLoading = useSelector(selectPortfolioLoading)

  useEffect(() => {
    if (!wallet || !foxy) return
    ;(async () => {
      setLoading(true)
      try {
        const chainAdapter = await chainAdapterManager.byChainId('eip155:1')
        const userAddress = await chainAdapter.getAddress({ wallet })
        const foxyOpportunities = await getFoxyOpportunities(balances, foxy, userAddress)

        // remove when Tokemak has api to get real apy
        for (const key in foxyOpportunities) {
          foxyOpportunities[key].apy = bnOrZero(getConfig().REACT_APP_FOXY_APY).toString()
        }

        setOpportunites(foxyOpportunities)
      } catch (error) {
        console.error('error', error)
      } finally {
        setLoading(false)
      }
    })()
  }, [wallet, foxyLoading, foxy, balances, balancesLoading, chainAdapterManager])

  const makeFiatAmount = useCallback(
    (opportunity: FoxyOpportunity) => {
      const asset = assets[opportunity.tokenAssetId]
      const pricePerShare = bnOrZero(opportunity.pricePerShare).div(`1e+${asset?.precision}`)
      const marketPrice = marketData[opportunity.tokenAssetId]?.price
      return bnOrZero(opportunity.balance)
        .div(`1e+${asset?.precision}`)
        .times(pricePerShare)
        .times(bnOrZero(marketPrice))
    },
    [assets, marketData],
  )

  const totalBalance = useMemo(
    () =>
      Object.values(opportunities).reduce((acc: BigNumber, opportunity: FoxyOpportunity) => {
        const amount = makeFiatAmount(opportunity)
        return acc.plus(bnOrZero(amount))
      }, bn(0)),
    [makeFiatAmount, opportunities],
  )

  const mergedOpportunities = useMemo(() => {
    return Object.values(opportunities).map(opportunity => {
      const asset = assets[opportunity.tokenAssetId]
      const fiatAmount = makeFiatAmount(opportunity)
      const marketPrice = marketData[opportunity.tokenAssetId]?.price ?? 0
      const tvl = bnOrZero(opportunity.tvl).div(`1e+${asset?.precision}`).times(marketPrice)
      const data = {
        ...opportunity,
        tvl,
        cryptoAmount: bnOrZero(opportunity.balance).div(`1e+${asset?.precision}`).toString(),
        fiatAmount: fiatAmount.toString(),
      }
      return data
    })
  }, [assets, makeFiatAmount, marketData, opportunities])

  return {
    opportunities: mergedOpportunities,
    loading: foxyLoading || loading,
    totalBalance: totalBalance.toString(),
  }
}
