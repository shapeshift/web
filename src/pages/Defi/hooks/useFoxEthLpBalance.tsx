import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
import { EarnOpportunityType } from 'features/defi/helpers/normalizeOpportunity'
import { UNISWAP_V2_WETH_FOX_POOL_ADDRESS } from 'features/defi/providers/fox-eth-lp/const'
import { useFoxEthLiquidityPool } from 'features/defi/providers/fox-eth-lp/hooks/useFoxEthLiquidityPool'
import { useLpApr } from 'plugins/foxPage/hooks/useLpApr'
import { useEffect, useState } from 'react'
import { useSelector } from 'react-redux'
import { useWallet } from 'hooks/useWallet/useWallet'
import { selectPortfolioAssetBalances, selectPortfolioLoading } from 'state/slices/selectors'

export type UseFoxEthLpBalanceReturn = {
  opportunity: EarnOpportunityType
  balance: string
  loading: boolean
}

const defaultOpportunity: EarnOpportunityType = {
  provider: DefiProvider.FoxEthLP,
  contractAddress: UNISWAP_V2_WETH_FOX_POOL_ADDRESS,
  rewardAddress: '',
  tvl: '',
  assetId: 'eip155:1/erc20:0x470e8de2ebaef52014a47cb5e6af86884947f08c',
  fiatAmount: '',
  cryptoAmount: '',
  chainId: 'eip155:1',
  isLoaded: false,
  type: DefiType.LiquidityPool,
}

export function useFoxEthLpBalance(): UseFoxEthLpBalanceReturn {
  const {
    state: { wallet },
  } = useWallet()
  const [loading, setLoading] = useState(false)
  const [opportunity, setOpportunity] = useState<EarnOpportunityType>(defaultOpportunity)
  const { calculateHoldings, getLpTVL } = useFoxEthLiquidityPool()
  const { lpApr } = useLpApr()

  const balances = useSelector(selectPortfolioAssetBalances)
  const balancesLoading = useSelector(selectPortfolioLoading)

  useEffect(() => {
    if (!wallet) return
    ;(async () => {
      setLoading(true)
      try {
        const lpBalance = await calculateHoldings()
        const totalSupply = await getLpTVL()
        setOpportunity({
          ...defaultOpportunity,
          cryptoAmount: lpBalance,
          isLoaded: true,
          apy: lpApr ?? undefined,
          tvl: totalSupply ?? '',
        })
      } catch (error) {
        console.error('error', error)
      } finally {
        setLoading(false)
      }
    })()
  }, [balances, wallet, balancesLoading, calculateHoldings, lpApr, getLpTVL])

  return {
    opportunity,
    balance: opportunity.cryptoAmount,
    loading: loading || balancesLoading,
  }
}
