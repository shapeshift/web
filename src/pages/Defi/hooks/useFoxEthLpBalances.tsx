import { ethAssetId, ethChainId } from '@shapeshiftoss/caip'
import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
import { EarnOpportunityType } from 'features/defi/helpers/normalizeOpportunity'
import {
  foxAssetId,
  foxEthLpAssetId,
  UNISWAP_V2_WETH_FOX_POOL_ADDRESS,
} from 'features/defi/providers/fox-eth-lp/constants'
import { useFoxEthLiquidityPool } from 'features/defi/providers/fox-eth-lp/hooks/useFoxEthLiquidityPool'
import { useLpApr } from 'plugins/foxPage/hooks/useLpApr'
import { useEffect, useState } from 'react'
import { useWallet } from 'hooks/useWallet/useWallet'
import { bnOrZero } from 'lib/bignumber/bignumber'
import { selectAssetById, selectMarketDataById } from 'state/slices/selectors'
import { useAppSelector } from 'state/store'

export type UseFoxEthLpBalancesReturn = {
  opportunity: EarnOpportunityType
  balance: string
  foxBalance: string | null
  ethBalance: string | null
  loading: boolean
}

const defaultOpportunity: EarnOpportunityType = {
  provider: DefiProvider.FoxEthLP,
  contractAddress: UNISWAP_V2_WETH_FOX_POOL_ADDRESS,
  rewardAddress: '',
  tvl: '',
  assetId: foxEthLpAssetId,
  fiatAmount: '',
  cryptoAmount: '',
  chainId: ethChainId,
  isLoaded: false,
  type: DefiType.LiquidityPool,
  icons: [
    'https://assets.coincap.io/assets/icons/eth@2x.png',
    'https://assets.coincap.io/assets/icons/fox@2x.png',
  ],
}

export function useFoxEthLpBalances(): UseFoxEthLpBalancesReturn {
  const {
    state: { wallet },
  } = useWallet()
  const [loading, setLoading] = useState<boolean>(true)
  const [foxBalance, setFoxBalance] = useState<string | null>(null)
  const [ethBalance, setEthBalance] = useState<string | null>(null)
  const [opportunity, setOpportunity] = useState<EarnOpportunityType>(defaultOpportunity)
  const { calculateHoldings, getLpTVL } = useFoxEthLiquidityPool()
  const { lpApr } = useLpApr()
  const foxMarketData = useAppSelector(state => selectMarketDataById(state, foxAssetId))
  const ethMarketData = useAppSelector(state => selectMarketDataById(state, ethAssetId))
  const lpAssetPrecision = useAppSelector(state =>
    selectAssetById(state, opportunity.assetId),
  ).precision

  useEffect(() => {
    if (!wallet) return
    ;(async () => {
      setLoading(true)
      try {
        const balances = await calculateHoldings()
        if (balances) {
          const { lpBalance, foxBalance, ethBalance } = balances
          setFoxBalance(foxBalance.toString())
          setEthBalance(ethBalance.toString())
          const totalLpBalance = bnOrZero(lpBalance).div(`1e${lpAssetPrecision}`)
          const ethValue = ethBalance.times(ethMarketData.price)
          const foxValue = foxBalance.times(foxMarketData.price)
          const fiatAmount = ethValue.plus(foxValue).toFixed(2)
          const totalSupply = await getLpTVL()
          setOpportunity({
            ...defaultOpportunity,
            cryptoAmount: totalLpBalance?.toString() ?? '0',
            fiatAmount,
            isLoaded: true,
            apy: lpApr ?? undefined,
            tvl: totalSupply ?? '',
          })
        }
      } catch (error) {
        console.error('error', error)
      } finally {
        setLoading(false)
      }
    })()
  }, [
    wallet,
    calculateHoldings,
    lpApr,
    getLpTVL,
    foxMarketData.price,
    ethMarketData.price,
    lpAssetPrecision,
  ])

  return {
    opportunity,
    balance: opportunity.cryptoAmount,
    loading: loading || !opportunity.apy,
    foxBalance,
    ethBalance,
  }
}
