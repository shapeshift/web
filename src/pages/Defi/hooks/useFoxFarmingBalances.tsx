import { ethAssetId, ethChainId } from '@shapeshiftoss/caip'
import { ChainAdapter } from '@shapeshiftoss/chain-adapters'
import { KnownChainIds } from '@shapeshiftoss/types'
import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
import { EarnOpportunityType } from 'features/defi/helpers/normalizeOpportunity'
import { foxEthLpAssetId } from 'features/defi/providers/fox-eth-lp/constants'
import { foxAssetId } from 'features/defi/providers/fox-eth-lp/constants'
import { getOpportunityData } from 'features/defi/providers/fox-farming/api'
import { FOX_FARMING_CONTRACT_ADDRESS } from 'features/defi/providers/fox-farming/constants'
import { FOX_TOKEN_CONTRACT_ADDRESS } from 'plugins/foxPage/const'
import { useLpApr } from 'plugins/foxPage/hooks/useLpApr'
import { useEffect, useMemo, useState } from 'react'
import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
import { useWallet } from 'hooks/useWallet/useWallet'
import { bnOrZero } from 'lib/bignumber/bignumber'
import { selectAssetById, selectMarketDataById } from 'state/slices/selectors'
import { useAppSelector } from 'state/store'

export type FoxFarmingEarnOpportunityType = {
  unclaimedRewards: string
} & EarnOpportunityType

export type UseFoxFarmingBalancesReturn = {
  opportunities: FoxFarmingEarnOpportunityType[]
  loading: boolean
  totalBalance: string
}

const defaultOpportunity: FoxFarmingEarnOpportunityType = {
  provider: DefiProvider.FoxFarming,
  contractAddress: FOX_FARMING_CONTRACT_ADDRESS,
  rewardAddress: FOX_TOKEN_CONTRACT_ADDRESS,
  tvl: '',
  assetId: foxEthLpAssetId,
  fiatAmount: '',
  cryptoAmount: '',
  unclaimedRewards: '',
  chainId: ethChainId,
  isLoaded: false,
  type: DefiType.Farming,
  icons: [
    'https://assets.coincap.io/assets/icons/eth@2x.png',
    'https://assets.coincap.io/assets/icons/fox@2x.png',
  ],
  opportunityName: 'Fox Farming V4',
}

export function useFoxFarmingBalances(): UseFoxFarmingBalancesReturn {
  const {
    state: { wallet },
  } = useWallet()
  const ethAsset = useAppSelector(state => selectAssetById(state, ethAssetId))

  const chainAdapterManager = getChainAdapterManager()
  const adapter = chainAdapterManager.get(ethAsset.chainId) as ChainAdapter<KnownChainIds>

  useEffect(() => {
    if (wallet && adapter) {
      ;(async () => {
        const address = await adapter.getAddress({ wallet })
        setConnectedWalletEthAddress(address)
      })()
    }
  }, [adapter, wallet])

  const [connectedWalletEthAddress, setConnectedWalletEthAddress] = useState<string | null>(null)
  const [loading, setLoading] = useState<boolean>(true)
  const [totalBalance, setTotalBalance] = useState<string>('')
  const [opportunities, setOpportunities] = useState<FoxFarmingEarnOpportunityType[]>([
    defaultOpportunity,
  ])
  const { isLpAprLoaded, lpApr } = useLpApr()
  const ethMarketData = useAppSelector(state => selectMarketDataById(state, ethAssetId))
  const foxMarketData = useAppSelector(state => selectMarketDataById(state, foxAssetId))
  const foxAssetPrecision = useAppSelector(state => selectAssetById(state, foxAssetId)).precision
  const lpAssetPrecision = useAppSelector(state =>
    selectAssetById(state, foxEthLpAssetId),
  ).precision

  useMemo(() => {
    if (!(wallet && connectedWalletEthAddress && lpApr)) return
    ;(async () => {
      setLoading(true)
      try {
        const newOpportunities = await Promise.all(
          opportunities.map(async opportunity => {
            const data = await getOpportunityData({
              contractAddress: opportunity.contractAddress,
              ethAssetPrecision: ethAsset.precision,
              ethPrice: ethMarketData.price,
              lpAssetPrecision,
              address: connectedWalletEthAddress,
              foxPrice: foxMarketData.price,
              foxAssetPrecision,
            })
            if (!data) return opportunity
            const { tvl, apr, balances, expired } = data
            return {
              ...opportunity,
              cryptoAmount: balances.cryptoBalance,
              fiatAmount: balances.fiatBalance,
              isLoaded: isLpAprLoaded,
              unclaimedRewards: balances.unclaimedRewards,
              expired,
              apy: lpApr
                ? bnOrZero(apr)
                    .plus(lpApr ?? 0)
                    .toString()
                : undefined,
              tvl,
            }
          }),
        )
        const totalOpBalances = newOpportunities.reduce(
          (acc, cur) => acc.plus(bnOrZero(cur.fiatAmount)),
          bnOrZero(0),
        )
        setTotalBalance(totalOpBalances.toFixed(2))
        setOpportunities(newOpportunities)
      } catch (error) {
        console.error('error', error)
      } finally {
        setLoading(false)
      }
    })()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [wallet, connectedWalletEthAddress, lpApr])

  return {
    opportunities,
    loading,
    totalBalance,
  }
}
