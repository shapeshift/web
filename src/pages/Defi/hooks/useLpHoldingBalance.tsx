import { bnOrZero } from 'lib/bignumber/bignumber'

import { useFoxEthLpBalances } from './useFoxEthLpBalances'
import { useFoxFarmingBalances } from './useFoxFarmingBalances'

export type useLpHoldingBalanceReturn = {
  lpHoldingsBalance: string
}

export function useLpHoldingBalance(): useLpHoldingBalanceReturn {
  const { opportunity: foxEthLpOpportunity } = useFoxEthLpBalances()
  const { totalBalance: foxFarmingTotalBalance } = useFoxFarmingBalances()

  // this should remain for the correct "netWorth" no matter what flags status are
  const lpHoldingsBalance = bnOrZero(foxEthLpOpportunity.fiatAmount).plus(foxFarmingTotalBalance)

  return {
    lpHoldingsBalance: lpHoldingsBalance.toString(),
  }
}
