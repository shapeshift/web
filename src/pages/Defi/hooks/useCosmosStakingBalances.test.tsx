import { cosmosAssetId } from '@shapeshiftoss/caip'
import { renderHook } from '@testing-library/react'
import type { PropsWithChildren } from 'react'
import { mockCosmosAccount, mockCosmosAccountWithStakingData } from 'test/mocks/accounts'
import { cosmos, mockAssetState } from 'test/mocks/assets'
import { mockMarketData } from 'test/mocks/marketData'
import { mockUpsertPortfolio } from 'test/mocks/portfolio'
import { MOCK_VALIDATORS } from 'test/mocks/validators'
import { TestProviders } from 'test/TestProviders'
import { useCosmosSdkStakingBalances } from 'pages/Defi/hooks/useCosmosSdkStakingBalances'
import { assets as assetsSlice } from 'state/slices/assetsSlice/assetsSlice'
import { marketData as marketDataSlice } from 'state/slices/marketDataSlice/marketDataSlice'
import { portfolio as portfolioSlice } from 'state/slices/portfolioSlice/portfolioSlice'
import { validatorData } from 'state/slices/validatorDataSlice/validatorDataSlice'
import { store } from 'state/store'

function setup() {
  const assetData = mockAssetState({
    byId: {
      [cosmos.assetId]: cosmos,
    },
    ids: [cosmos.assetId],
  })
  store.dispatch(assetsSlice.actions.setAssets(assetData))

  const cosmosMarketData = mockMarketData({ price: '77.55' })
  store.dispatch(
    marketDataSlice.actions.setCryptoMarketData({
      [cosmos.assetId]: cosmosMarketData,
    }),
  )

  const wrapper: React.FC<PropsWithChildren> = ({ children }) => (
    <TestProviders>{children}</TestProviders>
  )
  const { result } = renderHook(
    () =>
      useCosmosSdkStakingBalances({
        assetId: 'cosmos:cosmoshub-4/slip44:118',
      }),
    { wrapper },
  )
  return { result }
}

describe('useCosmosStakingBalances', () => {
  it('returns active and non active staking opportunities when staking and validators data are loaded', async () => {
    store.dispatch(
      validatorData.actions.upsertValidatorData({
        validators: MOCK_VALIDATORS,
      }),
    )

    const cosmosAccount = mockCosmosAccount(mockCosmosAccountWithStakingData)

    store.dispatch(
      portfolioSlice.actions.upsertPortfolio(mockUpsertPortfolio([cosmosAccount], [cosmosAssetId])),
    )

    const { result } = setup()
    expect(result.current.cosmosSdkStakingOpportunities).toMatchSnapshot()
    expect(result.current.totalBalance).toEqual('1.17')
  })
})
