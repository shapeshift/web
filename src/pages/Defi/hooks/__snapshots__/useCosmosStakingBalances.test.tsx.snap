// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCosmosStakingBalances returns active and non active staking opportunities when staking and validators data are loaded 1`] = `
Array [
  Object {
    "assetId": "cosmos:cosmoshub-4/slip44:118",
    "chainId": "cosmos:cosmoshub-4",
    "cryptoAmount": "0.000004",
    "fiatAmount": "0.00",
    "isLoaded": false,
    "rewards": "0",
    "totalDelegations": "4",
    "tvl": "0",
  },
  Object {
    "address": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "apr": "0.24",
    "assetId": "cosmos:cosmoshub-4/slip44:118",
    "chainId": "cosmos:cosmoshub-4",
    "commission": "0.100000000000000000",
    "cryptoAmount": "0.010115",
    "fiatAmount": "0.78",
    "isLoaded": true,
    "moniker": "Shapeshift DAO",
    "rewards": "3.831752143667562385",
    "tokens": "111111",
    "totalDelegations": "10115",
    "tvl": "8.61665805",
  },
  Object {
    "assetId": "cosmos:cosmoshub-4/slip44:118",
    "chainId": "cosmos:cosmoshub-4",
    "cryptoAmount": "0.005",
    "fiatAmount": "0.39",
    "isLoaded": false,
    "rewards": "12.688084635379675",
    "totalDelegations": "5000",
    "tvl": "0",
  },
]
`;

exports[`useCosmosStakingBalances returns empty array for active opportunities and the shapeshift validator as a staking opportunity when staking data is empty and validators data are loaded 1`] = `Array []`;
